### PR TITLE
GH-41089: [C++] Clean up remaining tasks related to half float casts

### PR DESCRIPTION
**Pull request by ClifHouck**
### Rationale for this change

Address remaining tasks brought up post-merge in this PR: https://github.com/apache/arrow/pull/40067

### What changes are included in this PR?

Adds a couple of tests and cleans up cruft on another test.

### Are these changes tested?

Yes, as these are test-only related changes.

### Are there any user-facing changes?

No.

* GitHub Issue: #41089

### CODE REVIEW

**Comment by ClifHouck** [comment]
@kou @pitrou Hopefully this addresses the remaining issues from #40067

**Comment by ClifHouck** [comment]
> Could you open a new issue for this?

Sure, filed https://github.com/apache/arrow/issues/41089

File: cpp/src/arrow/util/formatting_util_test.cc [hunk 1]
```diff
@@ -280,6 +282,32 @@ TEST(Formatting, Double) {
...
+TEST(Formatting, HalfFloat) {
+  StringFormatter<HalfFloatType> formatter;
+
+  AssertFormatting(formatter, Float16(0.0f).bits(), "0");
+  AssertFormatting(formatter, Float16(-0.0f).bits(), "-0");
+  AssertFormatting(formatter, Float16(1.5f).bits(), "1.5");
+
+  // Slightly adapted from values present here
+  // https://blogs.mathworks.com/cleve/2017/05/08/half-precision-16-bit-floating-point-arithmetic/
+  AssertFormatting(formatter, Float16::FromBits(0x3c00).bits(), "1");
```
Comment (pitrou) [inline cpp/src/arrow/util/formatting_util_test.cc:294]: This can certainly be simplified to `0x3c00`, no?

Comment (ClifHouck) [reply cpp/src/arrow/util/formatting_util_test.cc]: Good point, simplified.

**Review by pitrou — APPROVED** [review]
Thanks for the update @ClifHouck . This looks good to me.

**Pull request merged**